flytekit/core/type_engine.py

```diff
@@ -1419,6 +1428,226 @@ def guess_python_type(self, literal_type: LiteralType) -> list:  # type: ignore
+    Transformer that handles non-univariate tuple and NamedTuple.
+
+    1. typing.Tuple[T1, T2, T3]: Support.
+    2. typing.Tuple[T, ...]: Not support, we should use typing.List[T] instead.
+    3. NamedTuple(NamedTupleType, [(field_name1, field_type1), (field_name2, field_type2)]): Support.
+    """
+
+    def __init__(self):
+        super().__init__("Typed tuple", tuple)
+
+    @staticmethod
+    def default_tuple_name() -> str:
+        return ""
+
+    @staticmethod
+    def default_field_name(index: int = 0) -> str:
+        return f"t{index}"
+
+    @staticmethod
+    def tuple_fields_name_generator(length: int) -> Generator[str, None, None]:
+        for x in range(0, length):
+            yield TupleTransformer.default_field_name(x)
+
+    @staticmethod
+    def get_tuple_fields_type(t: Type[T]) -> Tuple:
+        """
+        Get the Tuple fields types.
+        """
+        args = get_args(t)
+        if len(args) == 0:
+            raise ValueError("Could not get the args of the Tuple type.")
+        if Ellipsis in args:
+            raise ValueError(f"Univariate tuple type {t} is not supported. Use List instead.")
+        return args
+
+    @staticmethod
+    def get_namedtuple_fields_type(t: Type[T]) -> Dict[str, Type[T]]:
+        """
+        Get the NamedTuple fields types.
+        """
+        return t.__annotations__
+
+    def assert_type(self, python_type: Type[T], python_val: T) -> None:
+        pass
+
+    def get_literal_type(self, t: Type[T]) -> Optional[LiteralType]:
+        """
+        Univariate tuples are not supported in Flyte
+        """
+        if is_namedtuple(t):
+            tuple_name = t.__name__
+            field_types = {
+                k: TypeEngine.to_literal_type(v) for k, v in TupleTransformer.get_namedtuple_fields_type(t).items()
+            }
+            order = list(getattr(t, "_fields"))
+
+            return _type_models.LiteralType(
+                tuple_type=TupleType(tuple_name=tuple_name, order=order, fields=field_types)
+            )
+        else:
+            try:
+                tuple_name = TupleTransformer.default_tuple_name()
+                fields_type = TupleTransformer.get_tuple_fields_type(t)
+                order = list(TupleTransformer.tuple_fields_name_generator(len(fields_type)))
+                field_types = OrderedDict(zip(order, (TypeEngine.to_literal_type(x) for x in fields_type)))
+
+                return _type_models.LiteralType(
+                    tuple_type=TupleType(tuple_name=tuple_name, order=order, fields=field_types)
+                )
+            except Exception as e:
+                raise ValueError(f"Type of Tuple {t} is not supported, {e}")
+
+    def to_literal(
+        self,
+        ctx: FlyteContext,
+        python_val: T,
+        python_type: Type[T],
+        expected: LiteralType,
+    ) -> Literal:
+        if expected.tuple_type.tuple_name != TupleTransformer.default_tuple_name():
+            # We are expected a NamedTuple
```

Comment:
Do you think we need a separate transformer for NamedTuple? `NamedTupleTransformer`

Reply:
Also, I have tried it before (maybe you can see that in some commits in this Pull Request), and it is really hard to maintain that.